documentation/projects/proposals/ingestion_server_removal/20240328-implementation_plan_ingestion_server_removal.md

```diff
@@ -0,0 +1,586 @@
+ingestion server as well.
+
+Finally we will update the monorepo's root
+[`docker-compose.yml`](https://github.com/WordPress/openverse/blob/05ff48d05f2163104151c5589cf352a156bc6a97/docker-compose.yml)
+to add a new `catalog-indexer-worker` service. Its build context should point to
+the nested `data_refresh/indexer_worker` directory, and it should map the
+exposed port to enable the API to be reached by the catalog.
+
+When this work is complete, it should be possible to run `just catalog/ipython`
+and curl the new indexer worker. The existing ingestion-server and
+indexer-worker services are unaffected (it is still possible to run legacy data
+refreshes locally and in production).
+
+### Implement distributed reindexing locally
+
+In this step we will add tasks to the data refresh DAGs to orchestrate the
+distributed reindex. At the end of this step, it will be possible to run a
+distributed reindex locally. Because the infrastructure work to create the EC2
+instances is not complete, it can not be run on production yet. The following
+code can all be refactored from
+[`distributed_reindex_scheduler.py`](https://github.com/WordPress/openverse/blob/main/ingestion_server/ingestion_server/distributed_reindex_scheduler.py).
+
+- Use Airflow's
+  [EC2Hook#describe_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.describe_instances)
+  to get a list of internal URLs bound to the indexer workers for the
+  appropriate environment. In the local environment, this will instead return
+  the URL for the catalog-indexer-worker Docker container.
+- Use
+  [EC2Hook#start_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.start_instances)
+  to start the appropriate EC2 instances. This step should skip in a local
+  environment.
+- Use dynamic task mapping to generate a Sensor for each of the expected indexer
+  workers, and ping its `healthcheck` endpoint until it passes.
+- Use dynamic task mapping to distribute reindexing across the indexer workers
+  by first calculating `start` and `end` indices that will split the records in
+  the media table into even portions, depending on the number of workers
+  available in the given environment. Then:
+  - POST to each worker's `reindexing_task` endpoint the `start_index` and
+    `end_index` it should handle
+  - Use a Sensor to ping the worker's `task/{task_id}` endpoint until the task
+    is complete, logging the progress as it goes
+  - Use the
+    [EC2Hook#stop_instances](https://airflow.apache.org/docs/apache-airflow-providers-amazon/stable/_api/airflow/providers/amazon/aws/hooks/ec2/index.html#airflow.providers.amazon.aws.hooks.ec2.EC2Hook.stop_instances)
+    to shut the instance down. This step should skip in a local environment.
```

Comment:
It might be worth noting here that we'll want this to happen regardless of upstream success or failures, so we're not keeping resources alive unnecessarily (the `none_failed` trigger rule, I believe).